### ISSUE TITLE
TensorFlow element-wise partial derivative over a batch of inputs

### ISSUE THREAD

**Issue by smao-astro** [lululxvi/deepxde#202]
Hi Lu,

For current implementation of partial derivative of outputs w.r.t. inputs, DeepXDE is using `tf.gradients`, see
https://github.com/lululxvi/deepxde/blob/ecfe9ba2fe391185386262be23ef27d0979ef343/deepxde/gradients.py#L36-L37
 However I am afraid this function was originally NOT implemented for "element-wise gradient over a batch".

For example, for 1D ODE (inputs: [N, 1], outputs: [N, 1], N the batch size), for i in N, the output w.r.t. `x_i` is `y_i`, and PINN need d(y_i)/d(x_i) in the loss function. According to [tf.gradients](https://www.tensorflow.org/api_docs/python/tf/gradients),
> It returns a list of Tensor of length len(xs) where each tensor is the sum(dy/dx) for y in ys and for x in XS.

As a result, I suspect if use `tf.gradients` to calculate the dy/dx, there are two problem:

1. Since `d(y_i)/d(x_j)` is needed only when `i=j`, the computation should only require O(N) computation, but if look at the docs, seems that `d(y_i)/d(x_j)` is evaluated for every i,j, so the code do N * N times evaluation of gradients.
2. What's worse, in some case such as there is a batch-normalization layer, `d(y_i)/d(x_j)` actually won't evaluate to 0 for `i != j`, and I suspect `tf.gradients` won't give the expected value.

Please have a look at it and confirm me, thanks.

**Comment by smao-astro** [comment]
All right, seems what I said above is not accurate, but I am still getting confused, see the code below

```python
import deepxde as dde
import numpy as np
from deepxde.backend import tf

x = tf.placeholder(dtype=tf.float32, shape=[3, 1])

y = tf.fill(dims=[3, 1], value=x[1, 0])

dydx_1 = tf.gradients(y,x)[0]

x_ = np.random.default_rng().random((3,1))

with tf.Session() as sess:
    print("x", sess.run(x, feed_dict={x: x_}))
    print("y", sess.run(y, feed_dict={x: x_}))
    print(sess.run(dydx_1, feed_dict={x: x_}))
```
with output
```x [[0.60492516]
 [0.21276382]
 [0.20765768]]
y [[0.21276382]
 [0.21276382]
 [0.21276382]]
[[0.]
 [3.]
 [0.]]
```

Any idea what happened here? Thanks.

**Comment by lululxvi** [comment]
`tf.fill` is slightly wired. What about other math functions?

**Comment by smao-astro** [comment]
```python
y2 = x - tf.reduce_mean(x, axis=0, keepdims=True)
dydx_2 = tf.gradients(y2,x)[0]

with tf.Session() as sess:
    print("x", sess.run(x, feed_dict={x: x_}))
    print("y", sess.run(y2, feed_dict={x: x_}))
    print(sess.run(dydx_2, feed_dict={x: x_}))
```
get
![Screen Shot 2021-01-20 at 10 08 44](https://user-images.githubusercontent.com/54871380/105117495-86d5f680-5b07-11eb-8042-b7a64e45de26.png)

I am expecting [2/3, 2/3, 2/3] here.

The only explanation I can give is that tf.gradients is evaluating N*N jacobian even only the diagonal elements is needed.

**Comment by lululxvi** [comment]
The result is 0 because tf.gradients sums all the y, which is 0 in the code, and then compute the gradients, which is again 0. In the document: "where each tensor is the sum(dy/dx) for y in ys".

If the function from x to y is a network, and the first dimension of x is the batch, i.e., y[i] is only a function of x[i], then there is no problem.

**Comment by Enowyz** [comment]
Hi, Lululxvi,

I tried to use `tf.where` and `tf.session` to generate a boolean condition for one of 2 constant values as a coefficient in my pde.
       
```python
        Cp_result = tf.where(x[:, 0:1] > 1, tf.zeros(x[:, 1:2].shape)+2973, tf.zeros(x[:, 1:2].shape)+3)
        Cp_result_cons = tf.constant(Cp_result)
        Cp = tf.Session().run(Cp_result_cons)
```   

I got this error:
**ValueError: Cannot convert a partially known TensorShape to a Tensor: (?, 1)**

Is there any way I can fix this? I am not familiar with placeholder, is it necessary to use it or there is another way to solve my error? 

Thank you.
